feat(search): Add indexes joining

BagritsevichStepan opened the pull request:
Adds indexes joining based on hash algorithm. Later we will integrate this into the `FT.AGGREGATE` command.

~This PR includes #5592 and #5603 . So, real number of lines is not so big :smile:~

dranikpg reviewed (dismissed):
I think by adding index and field names to the game you made your implementation a lot harder. Just by using some preprocessing (not on the data, but on the meta, read O(1)) you can you can simplify your code and get rid of many (costly) conversions. You can also simplify your containers - suddenly there was no need for all the custom wrappers.

I'll approve and its for you to decide if to go ahead if the strings - I'd personally not

BagritsevichStepan commented:
In the previous PR, I implemented it using indexes, but then switched to strings, thinking it would simplify the logic. However that introduced unnecessary overhead. I have reverted it back, so you can review the changes. MapsList is no longer needed, at least for the join logic.

Also, instead of relying on key names, we decided to use a pair of (shard_id, doc_id). From this information, the key name can be derived when needed

dranikpg reviewed (approved):
Much simpler than before
